cmd/gateway-b2.go

```diff
@@ -0,0 +1,652 @@
+			Object:    largeFile.Name,
+			UploadID:  largeFile.ID,
+			Initiated: largeFile.Timestamp,
+		})
+	}
+	return lmi, nil
+}
+
+// NewMultipartUpload upload object in multiple parts, uses B2's LargeFile upload API.
+// Large files can range in size from 5MB to 10TB.
+// Each large file must consist of at least 2 parts, and all of the parts except the
+// last one must be at least 5MB in size. The last part must contain at least one byte.
+// For more information - https://www.backblaze.com/b2/docs/large_files.html
+func (l *b2Objects) NewMultipartUpload(bucket string, object string, metadata map[string]string) (string, error) {
+	var uploadID string
+	bkt, err := l.Bucket(bucket)
+	if err != nil {
+		return uploadID, err
+	}
+	contentType := metadata["content-type"]
+	delete(metadata, "content-type")
+	lf, err := bkt.StartLargeFile(l.ctx, object, contentType, metadata)
+	if err != nil {
+		return uploadID, b2ToObjectError(traceError(err), bucket, object)
+	}
+	return lf.ID, nil
+}
+
+// CopyObjectPart copy part of object to other bucket and object.
+func (l *b2Objects) CopyObjectPart(srcBucket string, srcObject string, destBucket string, destObject string,
+	uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
+	return PartInfo{}, traceError(NotImplemented{})
+}
+
+// PutObjectPart puts a part of object in bucket, uses B2's LargeFile upload API.
+func (l *b2Objects) PutObjectPart(bucket string, object string, uploadID string, partID int, data *HashReader) (pi PartInfo, err error) {
+	bkt, err := l.Bucket(bucket)
+	if err != nil {
+		return pi, err
+	}
+	fc, err := bkt.File(uploadID, object).CompileParts(0, nil).GetUploadPartURL(l.ctx)
+	if err != nil {
+		return pi, b2ToObjectError(traceError(err), bucket, object, uploadID)
+	}
+	hr := newHexDigitsEndReader(data, data.Size())
+	sha1, err := fc.UploadPart(l.ctx, hr, "hex_digits_at_end", hr.Len(), partID)
+	if err != nil {
+		return pi, b2ToObjectError(traceError(err), bucket, object, uploadID)
+	}
+	if err = data.Verify(); err != nil { // Verify sha256sum after close.
+		return pi, b2ToObjectError(traceError(err), bucket, object, uploadID)
+	}
+	pi = PartInfo{
+		PartNumber:   partID,
+		LastModified: UTCNow(),
+		ETag:         sha1,
+		Size:         data.Size(),
+	}
+	return pi, nil
+}
+
+// ListObjectParts returns all object parts for specified object in specified bucket, uses B2's LargeFile upload API.
+func (l *b2Objects) ListObjectParts(bucket string, object string, uploadID string, partNumberMarker int, maxParts int) (lpi ListPartsInfo, err error) {
+	bkt, err := l.Bucket(bucket)
+	if err != nil {
+		return lpi, err
+	}
+	lpi = ListPartsInfo{
+		Bucket:           bucket,
+		Object:           object,
+		UploadID:         uploadID,
+		MaxParts:         maxParts,
+		PartNumberMarker: partNumberMarker,
+	}
+	// startPartNumber must be in the range 1 - 10000 for B2.
+	if partNumberMarker == 0 {
+		partNumberMarker = 1
```

Comment:
this will cause the list response to always skip the first part

Reply:
It won't @krishnasrinivas . Not in case of Backblaze.

Reply:
startPartNumber includes 1 and later, it is not the same style as S3.

Reply:
in that case we need to do partNumberMarker++ always because:

```
krishna@escape:~/dev/minio-tile-2$ miniodebug multipart listparts --bucket krishna1980 --object testobject --uploadid 4_z1efea6cee041ea395aeb021f_f205a55e6b4215c82_d20171003_m202115_c001_v0001093_t0023 --partmarker 2
{
  "Bucket": "krishna1980",
  "Key": "testobject",
  "UploadID": "4_z1efea6cee041ea395aeb021f_f205a55e6b4215c82_d20171003_m202115_c001_v0001093_t0023",
  "Initiator": {
    "ID": "02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4",
    "DisplayName": ""
  },
  "Owner": {
    "DisplayName": "",
    "ID": "02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4"
  },
  "StorageClass": "STANDARD",
  "PartNumberMarker": 2,
  "NextPartNumberMarker": 0,
  "MaxParts": 1000,
  "IsTruncated": false,
  "ObjectParts": [
    {
      "PartNumber": 2,
      "LastModified": "0001-01-01T00:00:00Z",
      "ETag": "\"5da54579c335464b6ba52090bf57d063384bb1e9\"",
      "Size": 2461
    },
    {
      "PartNumber": 3,
      "LastModified": "0001-01-01T00:00:00Z",
      "ETag": "\"5da54579c335464b6ba52090bf57d063384bb1e9\"",
      "Size": 2461
    }
  ],
  "EncodingType": ""
}
```

Reply:
Oh yes you are right. @krishnasrinivas